### ISSUE TITLE
Add locking to the FileStreamingTarget

### ISSUE THREAD

**Issue by mbrit** [onovotny/MetroLog#7]

**Comment by Sohra** [comment]
I've been getting a similar problem in the .NET 4.6 platform.  Even though both the WinRT & NetFx StreamingFileTarget class both ultimately subclass FileTargetBase that has locking in WriteAsyncCore method, I get an IOException thrown in FileTarget::DoWriteAsync on construction of the StreamWriter due to the log file being locked.  When using the ILogger interface, this results in an unobserved exception since all those methods just fire-and-forget the corresponding async method (it's not returned to the caller, and has no exception handling performed on it).

I've switched to using the ILoggerAsync interface, and chain a ContinueWith(, TaskContinuationOptions.OnlyOnFaulted) to the returned tasks to handle any exceptions that propagate up during logging.

I haven't figured out where, but I can only deduce there must be some kind of bug in the MetroLog.Internal.AsyncLock class that's allowing multiple calls (perhaps from multiple loggers with the same configuration) through to construct a StreamWriter for the same file.

In addition to fixing that bug, I propose that the MetroLog.Internal.Logger implementation of ILogger methods should observe any exceptions on the tasks they generate, and that the ILogManager interface be extended with a set of ILoggerAsync GetAsyncLogger methods to parallel the ILogger GetLogger methods.

**Comment by Sohra** [comment]
The IOException I experienced turned out to be my error.  I created two instances of the StreamingFileTarget configured with the same PathUnderAppData - one for the default configuration, and one for a particular logger I created with a custom configuration (which was intended a duplicate of the default with additional targets).  That meant two different locks for the same file.  I suppose I could avoid the exception by passing the default config's instance of the file target to my custom configuration, so that the AsyncLock would block the other logger's calls.

It would be really convenient if a FileTarget instance could obtain some kind of static lock on its file URI before creating the StreamWriter to prevent other instances writing to the same file at the same time.  It's not obvious to the developer that having different configurations that share a desire to log to the same file would compromise thread-safety like this, but I acknowledge my folly. 

The ability to enumerate targets in a given configuration (such as the default configuration) would facilitate copying one existing target into another easily (:
